activerecord/lib/active_record/connection_adapters/postgresql/oid/range.rb

```diff
@@ -69,13 +69,30 @@ def type_cast_single_for_database(value)
             def extract_bounds(value)
               from, to = value[1..-2].split(",", 2)
               {
-                from:          (from == "" || from == "-infinity") ? infinity(negative: true) : from,
-                to:            (to == "" || to == "infinity") ? infinity : to,
+                from:          (from == "" || from == "-infinity") ? infinity(negative: true) : unquote(from),
+                to:            (to == "" || to == "infinity") ? infinity : unquote(to),
                 exclude_start: value.start_with?("("),
                 exclude_end:   value.end_with?(")")
               }
             end
 
+            # When formatting the bound values of range types, PostgreSQL quotes
+            # the bound value using double-quotes in certain conditions (manual
+            # sections 8.17.5, 8.16.6). Within a double-quoted string, literal "
+            # and \ characters are themselves escaped. In input, PostgreSQL
+            # accepts multiple escape styles for " (either \" or "") but in
+            # output always uses "" (see 8.16.6).
+            def unquote(value)
+              if value.start_with?('"') && value.end_with?('"')
+                unquoted_value = value[1..-2]
+                unquoted_value.gsub!('""', '"')
+                unquoted_value.gsub!('\\\\', '\\')
+                unquoted_value
```

Comment:
Is this part still necessary? Regardless of unquoting the quoted characters or not, that doesn't work for now.

Reply:
Well, it's strictly still part of Postgres' output spec, so it seems wrong to leave it out just because Rails doesn't yet have the corresponding serialization side.

It also does work, for deserialization from Postgres: if there's a string range value `"ca\\t"..."do\"g"` in the database, Postgres will output that as `["ca""t","do\\g")` and Rails will correctly parse that with this code.

It's just that without the code that we split off into #39585, Rails is still lacking a way to perform the other direction and send such a value _to_ Postgres.

Reply:
Basically Active Record doesn't fully support custom type.
If you'd like to leave it for custom type, at least I'd not prefer untested change (i.e. all behavior change should be covered by tests).

Reply:
I think the problem is that this escaping handling handling is not specifically for custom types: it's about meeting Postgres' [specification for ranges](https://www.postgresql.org/docs/12/rangetypes.html#RANGETYPES-IO). Their documentation is clear on the cases where Postgres escapes the output and requires escaping on the input. It's actually only a coincidence that the formatting for the default range types happens to only include whitespace (and so not require input escaping).

I am happy to write a specific test that exercises this behavior. But the test for it will have to be a bit awkward, because we have chosen to hold back for now from correctly _serializing_ range values: it will have to only test the parsing. For the test to be nicer, we would have to reintroduce correcting the serialization to the PR.

Reply:
Done: I've written a test for the parsing behavior.